File: 11-container-with-most-water/container-with-most-water.ts
function maxArea(height: number[]): number {
    // area = min(l, r) * (r - l)
    // use a two pointer
    // if r > l, move left = r, r++
    // check if r 

    let l = 0
    let r = height.length - 1
    let max = 0

    while (l <= r) {
        let sum = Math.min(height[l], height[r]) * (r - l)
        max = Math.max(sum, max)

        if(height[r] > height[l]) {
            l++
        } else {
            r--
        }
    }

    return max
};

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # two pointer solution
        # calc area = min(l, r) **2
        # move smaller pointer

        area = 0

        l, r = 0, len(height) -1 

        while l<= r:
            curr_area = min(height[l], height[r]) * (r - l)
            area = max(area, curr_area)

            if height[l] > height[r]:
                r -= 1
            else:
                l+= 1

        return area